eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/jpa/LegacyJpaEventStorageEngine.java

```diff
@@ -0,0 +1,653 @@
+        if (isAggregateEvent) {
+            var nextSequence = aggregateSequencer.resolveBy(aggregateIdentifier).incrementAndGet();
+            return new GenericDomainEventMessage<>(
+                    aggregateType,
+                    aggregateIdentifier,
+                    nextSequence,
+                    event.getIdentifier(),
+                    event.name(),
+                    event.getPayload(),
+                    event.getMetaData(),
+                    event.getTimestamp()
+            );
+        } else {
+            // returns non-aggregate event, so the sequence is always 0
+            return new GenericDomainEventMessage<>(null,
+                                                   event.getIdentifier(),
+                                                   0L,
+                                                   event,
+                                                   event::getTimestamp);
+        }
+    }
+
+    @Override
+    public MessageStream<EventMessage<?>> source(@Nonnull SourcingCondition condition) {
+        var allCriteriaStream = condition
+                .criteria()
+                .stream()
+                .map(criteria -> this.eventsForCriteria(condition, criteria))
+                .reduce(MessageStream.empty(), MessageStream::concatWith);
+
+        var consistencyMarker = new AtomicReference<ConsistencyMarker>();
+        return allCriteriaStream.map(e -> {
+            var newMarker = consistencyMarker
+                    .accumulateAndGet(
+                            e.getResource(ConsistencyMarker.RESOURCE_KEY),
+                            (m1, m2) -> m1 == null ? m2 : m1.upperBound(m2)
+                    );
+            return e.withResource(ConsistencyMarker.RESOURCE_KEY, newMarker);
+        });
+    }
+
+    private MessageStream<EventMessage<?>> eventsForCriteria(SourcingCondition condition,
+                                                             EventCriteria criterion) {
+        var aggregateIdentifier = resolveAggregateIdentifier(criterion.tags());
+        var events = batchingOperations.readEventData(
+                aggregateIdentifier,
+                condition.start(),
+                condition.end()
+        );
+        var deserialized = upcastAndDeserializeDomainEvents(events, eventSerializer, upcasterChain);
+        return MessageStream.fromStream(
+                deserialized.asStream(),
+                e -> e,
+                LegacyJpaEventStorageEngine::fillContextWith
+        );
+    }
+
+    @Override
+    public MessageStream<EventMessage<?>> stream(@Nonnull StreamingCondition condition) {
+        var trackingToken = tokenOperations.assertGapAwareTrackingToken(condition.position());
+        var events = batchingOperations.readEventData(trackingToken);
+        var deserialized = upcastAndDeserializeTrackedEvents(events, eventSerializer, upcasterChain);
```

Comment:
We should not upcast, nor deserialize event messages. We should just create messages with exactly the format as the payload comes out of the database. Deserialization should happen on the consumer side, as each handler may have a different target representation to deserialize to.